### PR TITLE
ngtsc: rework compliance tests

### PR DESCRIPTION
**Pull request by petebacondarwin**
Supercedes #39582

This is the initial work to convert the compiler compliance tests to use file-based test cases, and also to test/update golden files for partial compilation.

The aim is to land this PR and then iterate - adding more test cases and improving the bazel rules.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -753,11 +753,11 @@ jobs:
       - setup_win
       - run:
           name: Build all windows CI targets
-          command: yarn bazel build --build_tag_filters=-ivy-only //packages/compiler-cli/... //tools/ts-api-guardian/...
+          command: yarn bazel build --build_tag_filters=-ivy-only-no-windows, //packages/compiler-cli/... //tools/ts-api-guardian/...
```
Comment (petebacondarwin) [inline .circleci/config.yml:756]: should there be a comma here?

```suggestion
          command: yarn bazel build --build_tag_filters=-ivy-only,-no-windows, //packages/compiler-cli/... //tools/ts-api-guardian/...
```

Comment (petebacondarwin) [reply .circleci/config.yml]: I fixed this - the comma was needed, but also we needed a `--build-tag-filters` option in the test command as well as the `--test_tag_filters` option.

**Comment by petebacondarwin** [comment]
@googlebot I consent.

**Review by josephperrott — APPROVED** [review]
LGTM

Reviewed-for: dev-infra

File: packages/compiler-cli/test/compliance/test_helpers/get_compliance_tests.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  }
+}
+
+/**
+ * Extract all the compliance tests from the TEST_CASES.json file at the `testConfigPath`.
+ *
+ * @param testConfigPath The path, relative to the `test_cases` basePath, of the `TEST_CASES.json`
+ *     config file.
+ */
+export function* getComplianceTests(testConfigPath: string): Generator<ComplianceTest> {
```
Comment (alxhub) [inline packages/compiler-cli/test/compliance/test_helpers/get_compliance_tests.ts:31]: Fancy! This function has bling `*`

**Review by alxhub — APPROVED** [review]
Fantastic work!

The only thing I would ask is that the`test/compliance` tree have a `README.md` which describes common scenarios - how do I focus a test? debug a test? regenerate a golden file? generate the goldens for a new test for the first time? etc.

**Comment by petebacondarwin** [comment]
> The only thing I would ask is that thetest/compliance tree have a README.md

@alxhub  - thanks for the review. The README is in the test_cases folder, but I can move it up a level.

File: packages/compiler-cli/test/compliance/test_helpers/get_compliance_tests.ts (hunk 2 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_helpers/get_compliance_tests.ts:31]: 🚀

**Pull request closed without merging**